cranelift-reader/src/parser.rs

```diff
@@ -2442,6 +2425,36 @@ impl<'a> Parser<'a> {
                 let lane = self.match_uimm8("expected lane number")?;
                 InstructionData::ExtractLane { opcode, lane, arg }
             }
+            InstructionFormat::Const => match explicit_control_type {
+                None => {
+                    return err!(
+                        self.loc,
+                        "Expected {:?} to have a controlling type variable, e.g. inst.i32x4",
+                        opcode
+                    )
+                }
+                Some(controlling_type) => {
+                    let uimm128 = self.match_uimm128_or_literals(controlling_type)?;
```

Comment:
In essence `parser.rs` uses the `FromStr` implementation and `write.rs` uses the `Display` implementation of whatever type we pick to represent constant values, so I think my comment on https://github.com/CraneStation/cranelift/pull/974#discussion_r324291764 applies here as well. I agree that it would be more flexible to perhaps parse to `ConstantData` here. We do need to constrain the parsing to the right number of bytes (i.e. 16 here) but this need not prevent us from using a more general type.